laundry/views.py

```diff
@@ -0,0 +1,160 @@
+        for week in range(1, 4):
+            # new_start is beginning of day, new_end is 27 hours after start
+            new_start = start - datetime.timedelta(weeks=week)
+            new_end = new_start + datetime.timedelta(hours=27)
+
+            new_snapshots = LaundrySnapshot.objects.filter(
+                hall_id=hall_id, date__gt=new_start, date__lte=new_end
+            )
+            snapshots = snapshots.union(new_snapshots)
+
+        return snapshots.order_by("-date")
+
+    def get(self, request, hall_id):
+
+        snapshots = self.get_queryset(hall_id)
+
+        # [0]: available washers, [1]: available dryers, [2]: total number of LaundrySnapshots
+        data = [(0, 0, 0)] * 27
+
+        # used calculate the start and end dates
+        min_date = timezone.localtime()
+        max_date = timezone.localtime() - datetime.timedelta(days=30)
+
+        for snapshot in snapshots:
+            date = snapshot.date.astimezone()
+
+            if date < min_date:
+                min_date = date
+
+            if date > max_date:
+                max_date = date
+
+            hour = date.hour
+
+            # accounts for the 3 hours on the next day
+            if (
+                calendar.day_name[timezone.localtime().weekday()]
+                == calendar.day_name[(date - datetime.timedelta(days=1)).weekday()]
+            ):
+                hour = date.hour + 24
+
+            # adds total number of available washers and dryers
+            data[hour] = (
+                data[hour][0] + snapshot.available_washers,
+                data[hour][1] + snapshot.available_dryers,
+                data[hour][2] + 1,
+            )
+
+        # collects total washers and dryers
+        try:
+            total_washers = snapshots.first().total_washers
+            total_dryers = snapshots.first().total_dryers
+        except AttributeError:
+            total_washers = 0
+            total_dryers = 0
+
+        return Response(
+            {
+                "hall_name": laundry.id_to_hall[int(hall_id)],
+                "location": laundry.id_to_location[int(hall_id)],
+                "day_of_week": calendar.day_name[timezone.localtime().weekday()],
+                "start_date": min_date.date(),
+                "end_date": max_date.date(),
+                "washer_data": {x: self.safe_division(data[x][0], data[x][2]) for x in range(27)},
+                "dryer_data": {x: self.safe_division(data[x][1], data[x][2]) for x in range(27)},
+                "total_number_of_washers": total_washers,
+                "total_number_of_dryers": total_dryers,
+            }
+        )
+
+
+class Preferences(APIView):
+    def get(self, request):
+
+        # tests if user is logged-in
+        if not request.user.is_authenticated:
+            return Response({"rooms": []})
+
+        preferences = Profile.objects.get(user=request.user).preferences.all()
+
+        hall_ids = [x.hall_id for x in preferences]
```

Comment:
You can combine this into a single `values_list` (I think that's the function) from a queryset